src/language-js/needs-parens.js

```diff
@@ -16,37 +16,95 @@ function hasClosureCompilerTypeCastComment(text, path) {
   // Syntax example: var x = /** @type {string} */ (fruit);
 
   const n = path.getValue();
+  const parenStart = getParenStart(n);
 
   return (
-    isParenthesized(n) &&
-    (hasTypeCastComment(n) || hasAncestorTypeCastComment(0))
+    parenStart !== -1 &&
+    (getTypeCastIndex(n, parenStart, -1) !== -1 ||
+      hasAncestorTypeCastComment(0, parenStart))
   );
 
   // for sub-item: /** @type {array} */ (numberOrString).map(x => x);
-  function hasAncestorTypeCastComment(index) {
+  function hasAncestorTypeCastComment(index, parenStart) {
     const ancestor = path.getParentNode(index);
-    return ancestor && !isParenthesized(ancestor)
-      ? hasTypeCastComment(ancestor) || hasAncestorTypeCastComment(index + 1)
-      : false;
+    if (!ancestor) {
+      return false;
+    }
+
+    const ancestorParenStart = getParenStart(ancestor);
+    const typeCastIndex = getTypeCastIndex(
+      ancestor,
+      parenStart,
+      ancestorParenStart
+    );
+
+    // The node can inherit a typecast all the way to the root expression,
+    // provided some ancestor is not wrapped in parens. If so, then any
+    // typecast above that ancestor applies to the ancestor, not our node.
+    if (typeCastIndex === -1) {
+      return (
+        ancestorParenStart === -1 &&
+        hasAncestorTypeCastComment(index + 1, parenStart)
+      );
+    }
+
+    // There's an annoying issue if both our node contains a typecast, and this
+    // ancestor contains a typecast. Unfortunately, both comments are applied
+    // to ancestor. So we need to detect this situation, and insert hack paren
+    // comments to simulate the correct wrapping.
+    if (
+      ancestorParenStart !== -1 &&
+      getTypeCastIndex(ancestor, ancestorParenStart, -1) !== -1
+    ) {
+      ancestor.comments.splice(
+        typeCastIndex,
+        0,
+        comments.makeOpenParenComment()
+      );
+      n.comments = n.comments || [];
+      n.comments.unshift(comments.makeCloseParenComment());
+      return false;
+    }
+    return true;
   }
 
-  function hasTypeCastComment(node) {
-    return (
-      node.comments &&
-      node.comments.some(
-        comment =>
-          comment.leading &&
-          comments.isBlockComment(comment) &&
-          isTypeCastComment(comment.value)
-      )
+  function getTypeCastIndex(node, parenStart, ancestorParenStart) {
+    if (!node.comments) {
+      return -1;
+    }
+    return node.comments.findIndex(
+      comment =>
+        comment.leading &&
+        comment.end <= parenStart &&
+        comment.start >= ancestorParenStart &&
+        comments.isBlockComment(comment) &&
+        isTypeCastComment(comment.value)
     );
   }
 
-  function isParenthesized(node) {
+  function getParenStart(node) {
     // Closure typecast comments only really make sense when _not_ using
     // typescript or flow parsers, so we take advantage of the babel parser's
     // parenthesized expressions.
```

Comment:
I forgot this feature was Babel-only. We should use Babel's `createParenthesizedExpressions` option then. This should give us nested parens in the AST. Could you look into that?